package sayitobar;

import java.awt.Toolkit;
import java.awt.event.WindowEvent;
import java.io.File;
import javax.swing.JFileChooser;
import javax.swing.filechooser.FileNameExtensionFilter;

/**
 *
 * @author mac
 */
public class DetectorMenu extends javax.swing.JFrame {

    /**
     * Creates new form MainMenu
     */
    public DetectorMenu() {
        initComponents();
        
        // Set outputFile location
        outPathLoc.setText(System.getProperty("user.home") + "/Desktop");
        
        // Window location
        setLocation(GenomeDetector.WinX - getWidth()/2, GenomeDetector.WinY - getHeight()/2);
        
        // Set jtextFields
        P_GenesCellValue.setText(Detector.P_GenesCellValue);
        L_GenesCellValue.setText(Detector.L_GenesCellValue);
        
        // Set splitters
        if (Detector.splitters.trim().equals(""))  Detector.splitters = "/|,|=|;";
        splitElementsTxtfld.setText(Detector.splitters.replace("|", " ").trim());
    }

    /**
     * This method is called from within the constructor to initialize the form.
     * WARNING: Do NOT modify this code. The content of this method is always
     * regenerated by the Form Editor.
     */
    @SuppressWarnings("unchecked")
    // <editor-fold defaultstate="collapsed" desc="Generated Code">//GEN-BEGIN:initComponents
    private void initComponents() {

        Title = new javax.swing.JLabel();
        searchButton = new javax.swing.JToggleButton();
        patientFileLoc = new javax.swing.JTextField();
        selectPatientButton = new javax.swing.JButton();
        selectGenePoolButton = new javax.swing.JButton();
        genePoolFileLoc = new javax.swing.JTextField();
        outFileName = new javax.swing.JTextField();
        outFileLabel = new javax.swing.JLabel();
        outPathLoc = new javax.swing.JTextField();
        selectOutputPathButton = new javax.swing.JButton();
        outPathLabel = new javax.swing.JLabel();
        patientFileLabel = new javax.swing.JLabel();
        genePoolFileLabel = new javax.swing.JLabel();
        infoButton = new javax.swing.JButton();
        jSeparator2 = new javax.swing.JSeparator();
        backButton = new javax.swing.JButton();
        P_GenesCellValue = new javax.swing.JTextField();
        description1 = new javax.swing.JLabel();
        description2 = new javax.swing.JLabel();
        L_GenesCellValue = new javax.swing.JTextField();
        splitElementsTxtfld = new javax.swing.JTextField();
        jLabel1 = new javax.swing.JLabel();
        splitCheckbox = new javax.swing.JCheckBox();
        devButton = new javax.swing.JButton();

        setDefaultCloseOperation(javax.swing.WindowConstants.DISPOSE_ON_CLOSE);
        setCursor(new java.awt.Cursor(java.awt.Cursor.DEFAULT_CURSOR));
        setResizable(false);

        Title.setFont(new java.awt.Font("Futura", 0, 24)); // NOI18N
        Title.setHorizontalAlignment(javax.swing.SwingConstants.CENTER);
        Title.setText("Gene Detector");

        searchButton.setText("Search for occurrences");
        searchButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                searchButtonActionPerformed(evt);
            }
        });

        patientFileLoc.setEditable(false);
        patientFileLoc.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                patientFileLocActionPerformed(evt);
            }
        });

        selectPatientButton.setIcon(javax.swing.UIManager.getIcon("FileChooser.directoryIcon"));
        selectPatientButton.setToolTipText("");
        selectPatientButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                selectPatientButtonActionPerformed(evt);
            }
        });

        selectGenePoolButton.setIcon(javax.swing.UIManager.getIcon("FileChooser.directoryIcon"));
        selectGenePoolButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                selectGenePoolButtonActionPerformed(evt);
            }
        });

        genePoolFileLoc.setEditable(false);
        genePoolFileLoc.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                genePoolFileLocActionPerformed(evt);
            }
        });

        outFileName.setToolTipText("");
        outFileName.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                outFileNameActionPerformed(evt);
            }
        });

        outFileLabel.setText("Name of output file (.xlsx):");

        outPathLoc.setEditable(false);
        outPathLoc.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                outPathLocActionPerformed(evt);
            }
        });

        selectOutputPathButton.setIcon(javax.swing.UIManager.getIcon("FileChooser.directoryIcon"));
        selectOutputPathButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                selectOutputPathButtonActionPerformed(evt);
            }
        });

        outPathLabel.setText("Select output folder (where to store it):");

        patientFileLabel.setText("Select first gene file (.xlsx):");

        genePoolFileLabel.setText("Select second gene file (.xlsx):");

        infoButton.setFont(new java.awt.Font("Helvetica Neue", 0, 10)); // NOI18N
        javax.swing.Icon icon = javax.swing.UIManager.getIcon("OptionPane.informationIcon");
        java.awt.image.BufferedImage bufferedImage = new java.awt.image.BufferedImage(icon.getIconWidth(), icon.getIconHeight(), java.awt.image.BufferedImage.TYPE_INT_ARGB);
        java.awt.Graphics2D g_ = bufferedImage.createGraphics();
        icon.paintIcon(null, g_, 0, 0);
        g_.dispose();
        javax.swing.Icon icon_ = new javax.swing.ImageIcon(bufferedImage.getScaledInstance(16, 16, java.awt.Image.SCALE_SMOOTH));
        infoButton.setIcon(icon_);
        infoButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                infoButtonActionPerformed(evt);
            }
        });

        backButton.setIcon(javax.swing.UIManager.getIcon("FileChooser.homeFolderIcon"));
        backButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                backButtonActionPerformed(evt);
            }
        });

        P_GenesCellValue.setText("-");
        P_GenesCellValue.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                P_GenesCellValueActionPerformed(evt);
            }
        });

        description1.setText("Column name of gene symbols (first file)");

        description2.setText("Column name of gene symbols (second file)");

        L_GenesCellValue.setText("-");
        L_GenesCellValue.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                L_GenesCellValueActionPerformed(evt);
            }
        });

        splitElementsTxtfld.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                splitElementsTxtfldActionPerformed(evt);
            }
        });

        jLabel1.setFont(new java.awt.Font("Helvetica Neue", 2, 12)); // NOI18N
        jLabel1.setText("(space between every splitter) ");

        splitCheckbox.setText("Split cell text with");
        splitCheckbox.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                splitCheckboxActionPerformed(evt);
            }
        });

        devButton.setFont(new java.awt.Font("Helvetica Neue", 0, 10)); // NOI18N
        devButton.setIcon(javax.swing.UIManager.getIcon("FileView.hardDriveIcon"));
        devButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                devButtonActionPerformed(evt);
            }
        });

        javax.swing.GroupLayout layout = new javax.swing.GroupLayout(getContentPane());
        getContentPane().setLayout(layout);
        layout.setHorizontalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addGap(182, 182, 182)
                .addComponent(searchButton, javax.swing.GroupLayout.PREFERRED_SIZE, 164, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addContainerGap(javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
            .addGroup(layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addGroup(layout.createSequentialGroup()
                        .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                            .addGroup(layout.createSequentialGroup()
                                .addComponent(selectOutputPathButton, javax.swing.GroupLayout.PREFERRED_SIZE, 26, javax.swing.GroupLayout.PREFERRED_SIZE)
                                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                                    .addGroup(layout.createSequentialGroup()
                                        .addComponent(outFileLabel)
                                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                                        .addComponent(outFileName, javax.swing.GroupLayout.PREFERRED_SIZE, 295, javax.swing.GroupLayout.PREFERRED_SIZE))
                                    .addComponent(outPathLabel)
                                    .addComponent(outPathLoc, javax.swing.GroupLayout.PREFERRED_SIZE, 454, javax.swing.GroupLayout.PREFERRED_SIZE)
                                    .addComponent(jSeparator2, javax.swing.GroupLayout.PREFERRED_SIZE, 454, javax.swing.GroupLayout.PREFERRED_SIZE))
                                .addGap(7, 7, 7))
                            .addGroup(layout.createSequentialGroup()
                                .addComponent(backButton, javax.swing.GroupLayout.PREFERRED_SIZE, 25, javax.swing.GroupLayout.PREFERRED_SIZE)
                                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                                .addComponent(Title, javax.swing.GroupLayout.PREFERRED_SIZE, 199, javax.swing.GroupLayout.PREFERRED_SIZE)
                                .addGap(105, 105, 105)
                                .addComponent(devButton, javax.swing.GroupLayout.PREFERRED_SIZE, 25, javax.swing.GroupLayout.PREFERRED_SIZE)
                                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)))
                        .addComponent(infoButton, javax.swing.GroupLayout.PREFERRED_SIZE, 24, javax.swing.GroupLayout.PREFERRED_SIZE))
                    .addGroup(layout.createSequentialGroup()
                        .addComponent(selectGenePoolButton, javax.swing.GroupLayout.PREFERRED_SIZE, 26, javax.swing.GroupLayout.PREFERRED_SIZE)
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                        .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING, false)
                            .addComponent(genePoolFileLabel)
                            .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.TRAILING)
                                .addComponent(genePoolFileLoc)
                                .addGroup(layout.createSequentialGroup()
                                    .addGap(0, 0, Short.MAX_VALUE)
                                    .addComponent(description2)
                                    .addGap(92, 92, 92)
                                    .addComponent(L_GenesCellValue, javax.swing.GroupLayout.PREFERRED_SIZE, 105, javax.swing.GroupLayout.PREFERRED_SIZE)))
                            .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.TRAILING)
                                .addComponent(jLabel1)
                                .addGroup(layout.createSequentialGroup()
                                    .addComponent(splitCheckbox)
                                    .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                                    .addComponent(splitElementsTxtfld, javax.swing.GroupLayout.PREFERRED_SIZE, 159, javax.swing.GroupLayout.PREFERRED_SIZE)))))
                    .addGroup(layout.createSequentialGroup()
                        .addComponent(selectPatientButton, javax.swing.GroupLayout.PREFERRED_SIZE, 25, javax.swing.GroupLayout.PREFERRED_SIZE)
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                        .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.TRAILING)
                            .addGroup(layout.createSequentialGroup()
                                .addComponent(description1, javax.swing.GroupLayout.PREFERRED_SIZE, 335, javax.swing.GroupLayout.PREFERRED_SIZE)
                                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                                .addComponent(P_GenesCellValue, javax.swing.GroupLayout.PREFERRED_SIZE, 107, javax.swing.GroupLayout.PREFERRED_SIZE))
                            .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                                .addComponent(patientFileLabel)
                                .addComponent(patientFileLoc, javax.swing.GroupLayout.PREFERRED_SIZE, 454, javax.swing.GroupLayout.PREFERRED_SIZE)))))
                .addContainerGap())
        );
        layout.setVerticalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addComponent(Title)
                    .addComponent(infoButton, javax.swing.GroupLayout.PREFERRED_SIZE, 24, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addComponent(backButton)
                    .addComponent(devButton, javax.swing.GroupLayout.PREFERRED_SIZE, 24, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addGap(34, 34, 34)
                .addComponent(patientFileLabel)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(selectPatientButton)
                    .addComponent(patientFileLoc, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(description1)
                    .addComponent(P_GenesCellValue, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addGap(28, 28, 28)
                .addComponent(genePoolFileLabel)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(selectGenePoolButton)
                    .addComponent(genePoolFileLoc, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(L_GenesCellValue, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addComponent(description2))
                .addGap(28, 28, 28)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(splitElementsTxtfld, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addComponent(splitCheckbox))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addComponent(jLabel1)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED, 45, Short.MAX_VALUE)
                .addComponent(outPathLabel)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addComponent(selectOutputPathButton)
                    .addComponent(outPathLoc, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addGap(24, 24, 24)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(outFileName, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addComponent(outFileLabel))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                .addComponent(jSeparator2, javax.swing.GroupLayout.PREFERRED_SIZE, 10, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addComponent(searchButton)
                .addGap(12, 12, 12))
        );

        pack();
    }// </editor-fold>//GEN-END:initComponents

    private void searchButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_searchButtonActionPerformed
        if ("".equals(patientFileLoc.getText()) || "".equals(genePoolFileLoc.getText()) || "".equals(outFileName.getText()) || "".equals(outPathLoc.getText())) {
            // if any places empty, do not run it.
            return;
        }

        Detector.P_GenesCellValue = P_GenesCellValue.getText();
        Detector.L_GenesCellValue = L_GenesCellValue.getText();
        String splitters          = splitElementsTxtfld.getText();
        for (int i=0; i < "\\.[]{}()<>*+-=!?^$|".length(); i++)  // add "\\" in front of special characters
            splitters = splitters.replace(String.valueOf("\\.[]{}()<>*+-=!?^$|".charAt(i)), "\\\\" + String.valueOf("\\.[]{}()<>*+-=!?^$|".charAt(i)));
        Detector.splitters      = splitters.replace(" ", "|");
        if (!splitCheckbox.isSelected())  Detector.splitters = "                                                                                     ";
        
        
        try {
            setCursor(new java.awt.Cursor(java.awt.Cursor.WAIT_CURSOR));
            
            Detector.start(
                    patientFileLoc.getText(),
                    genePoolFileLoc.getText(),
                    outPathLoc.getText(),
                    "/" + outFileName.getText()
            );
            
            setCursor(new java.awt.Cursor(java.awt.Cursor.DEFAULT_CURSOR));
            javax.swing.JOptionPane.showMessageDialog(this, "File successfully generated:\n" + outPathLoc.getText() + "/" + outFileName.getText(), "Success", 1);

        } catch (Exception e) {  // IOException doesn't work here, trust me
            setCursor(new java.awt.Cursor(java.awt.Cursor.DEFAULT_CURSOR));
            
            System.out.println("Couldn't start Detector.start() (line " + e.getStackTrace()[0].getLineNumber() + "): " + e);
            e.printStackTrace();
            
            javax.swing.JOptionPane.showMessageDialog(this, "Search failed: One of the files are faulty. (Gene Detector)\n\nPlease click (?) on the right corner of the\nscreen to see the rules of input excel files. \nPlease check the formatting and try again.\n\n\nDebug:\n" + e, "Search failed", 2);
        }
        
        // dispose();
        // System.exit(0);  // exit on task accompilation
    }//GEN-LAST:event_searchButtonActionPerformed

    private void patientFileLocActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_patientFileLocActionPerformed
        // TODO add your handling code here:
    }//GEN-LAST:event_patientFileLocActionPerformed

    private void genePoolFileLocActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_genePoolFileLocActionPerformed
        // TODO add your handling code here:
    }//GEN-LAST:event_genePoolFileLocActionPerformed

    private void selectGenePoolButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_selectGenePoolButtonActionPerformed
        JFileChooser fileChooser = new JFileChooser();
        
        fileChooser.setCurrentDirectory(new File(System.getProperty("user.home") + "/Desktop/"));
        fileChooser.setFileFilter(new FileNameExtensionFilter("Excel File (2013-)", "xlsx"));
        fileChooser.setAcceptAllFileFilterUsed(false);
        fileChooser.setDialogTitle("Select desired gene pool file");

        int response = fileChooser.showDialog(this, "Select");
        
        if (response == JFileChooser.APPROVE_OPTION) {
            String filePath = fileChooser.getSelectedFile().getAbsolutePath();
            genePoolFileLoc.setText(filePath);
        }
    }//GEN-LAST:event_selectGenePoolButtonActionPerformed

    private void outFileNameActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_outFileNameActionPerformed
        // TODO add your handling code here:
    }//GEN-LAST:event_outFileNameActionPerformed

    private void outPathLocActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_outPathLocActionPerformed
        // TODO add your handling code here:
    }//GEN-LAST:event_outPathLocActionPerformed

    private void selectOutputPathButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_selectOutputPathButtonActionPerformed
        JFileChooser fileChooser = new JFileChooser();
        
        fileChooser.setCurrentDirectory(new File(System.getProperty("user.home") + "/Desktop/"));
        fileChooser.setFileSelectionMode(JFileChooser.DIRECTORIES_ONLY);
        fileChooser.setAcceptAllFileFilterUsed(false);
        fileChooser.setDialogTitle("Select output folder");

        int response = fileChooser.showDialog(this, "Select");
        
        if (response == JFileChooser.APPROVE_OPTION) {
            String filePath = fileChooser.getSelectedFile().getAbsolutePath();
            
            outPathLoc.setText(filePath);
        }

    }//GEN-LAST:event_selectOutputPathButtonActionPerformed

    private void selectPatientButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_selectPatientButtonActionPerformed
        JFileChooser fileChooser = new JFileChooser();
        
        fileChooser.setCurrentDirectory(new File(System.getProperty("user.home") + "/Desktop/"));
        fileChooser.setFileFilter(new FileNameExtensionFilter("Excel File (2013-)", "xlsx"));
        fileChooser.setAcceptAllFileFilterUsed(false);
        fileChooser.setDialogTitle("Select patient genes file");
        
        int response = fileChooser.showDialog(this, "Select");
        
        if (response == JFileChooser.APPROVE_OPTION) {
            String filePath = fileChooser.getSelectedFile().getAbsolutePath();
            
            patientFileLoc.setText(filePath);
            
            outFileName.setText("OUTPUT-" + filePath.split("/")[filePath.split("/").length-1]);
        }
    }//GEN-LAST:event_selectPatientButtonActionPerformed

    
    public void close() {
        WindowEvent closeWin = new WindowEvent(this, WindowEvent.WINDOW_CLOSING);
        Toolkit.getDefaultToolkit().getSystemEventQueue().postEvent(closeWin);
    }
     
    private void infoButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_infoButtonActionPerformed
        System.out.println("Opening info window");
        
        GenomeDetector.WinX = getX() + getWidth() /2;
        GenomeDetector.WinY = getY() + getHeight()/2;
        
        close();
        infoWindow.wakeup(infoWindow.Messages.DETECTOR);
    }//GEN-LAST:event_infoButtonActionPerformed

    private void backButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_backButtonActionPerformed
        // go to home screen
        System.out.println("Back to home menu");
        
        GenomeDetector.WinX = getX() + getWidth() /2;
        GenomeDetector.WinY = getY() + getHeight()/2;
        
        close();
        MainMenu.wakeup();
    }//GEN-LAST:event_backButtonActionPerformed

    private void P_GenesCellValueActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_P_GenesCellValueActionPerformed
        // TODO add your handling code here:
    }//GEN-LAST:event_P_GenesCellValueActionPerformed

    private void L_GenesCellValueActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_L_GenesCellValueActionPerformed
        // TODO add your handling code here:
    }//GEN-LAST:event_L_GenesCellValueActionPerformed

    private void splitElementsTxtfldActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_splitElementsTxtfldActionPerformed
        // TODO add your handling code here:
    }//GEN-LAST:event_splitElementsTxtfldActionPerformed

    private void splitCheckboxActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_splitCheckboxActionPerformed
        // TODO add your handling code here:
    }//GEN-LAST:event_splitCheckboxActionPerformed

    private void devButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_devButtonActionPerformed
        System.out.println("Opening DEV CONSOLE");

        Thread t = new Thread() {
            @Override
            public void run() {
                devConsole.wakeup();
            }
        };
        t.start();
    }//GEN-LAST:event_devButtonActionPerformed

    
    
    public static void wakeup() {
        try {
            for (javax.swing.UIManager.LookAndFeelInfo info : javax.swing.UIManager.getInstalledLookAndFeels()) {
                if ("Nimbus".equals(info.getName())) {
                    javax.swing.UIManager.setLookAndFeel(info.getClassName());
                    break;
                }
            }
        } catch (ClassNotFoundException | InstantiationException | IllegalAccessException | javax.swing.UnsupportedLookAndFeelException ex) {
            java.util.logging.Logger.getLogger(infoWindow.class.getName()).log(java.util.logging.Level.SEVERE, null, ex);
        }

        /* Create and display the form */
        java.awt.EventQueue.invokeLater(() -> {
            new DetectorMenu().setVisible(true);
        });
    }

    // Variables declaration - do not modify//GEN-BEGIN:variables
    private javax.swing.JTextField L_GenesCellValue;
    private javax.swing.JTextField P_GenesCellValue;
    private javax.swing.JLabel Title;
    private javax.swing.JButton backButton;
    private javax.swing.JLabel description1;
    private javax.swing.JLabel description2;
    private javax.swing.JButton devButton;
    private javax.swing.JLabel genePoolFileLabel;
    private javax.swing.JTextField genePoolFileLoc;
    private javax.swing.JButton infoButton;
    private javax.swing.JLabel jLabel1;
    private javax.swing.JSeparator jSeparator2;
    private javax.swing.JLabel outFileLabel;
    private javax.swing.JTextField outFileName;
    private javax.swing.JLabel outPathLabel;
    private javax.swing.JTextField outPathLoc;
    private javax.swing.JLabel patientFileLabel;
    private javax.swing.JTextField patientFileLoc;
    private javax.swing.JToggleButton searchButton;
    private javax.swing.JButton selectGenePoolButton;
    private javax.swing.JButton selectOutputPathButton;
    private javax.swing.JButton selectPatientButton;
    private javax.swing.JCheckBox splitCheckbox;
    private javax.swing.JTextField splitElementsTxtfld;
    // End of variables declaration//GEN-END:variables
}
